### DISCUSSION TITLE
Cant override field's value on a submission after submit

### DISCUSSION THREAD

**Discussion by palomamtnez** [solspace/craft-freeform#429] [Questions & Answers] [answered]
I need to override the value of a field from a submission of a form on an after-submit event
It looks like the value is not getting stored in the submission ; however it is sent to the admin email with the new value, but not stored

I have this event listener

```
Event::on(
    SubmissionsService::class,
    SubmissionsService::EVENT_AFTER_SUBMIT,
    function (SubmitEvent $event) {
        $form = $event->getForm();
        try {
            $myNewValue = getenv('CUSTOM_URL') . $event->getSubmission()->id;
            $form->getLayout()->getFieldByHandle('submissionUrl')->setValue($mynewValue);
```

so on my email, I get the new value I want:
  
![Screen Shot 2022-08-11 at 2 20 21 PM](https://user-images.githubusercontent.com/13962572/184054282-b3dd571f-76e2-4a9c-97c7-451159c62c17.png)

but on my stored submissions in the CMS it looks empty: 

![Screen Shot 2022-08-11 at 2 25 38 PM](https://user-images.githubusercontent.com/13962572/184054816-30be167d-cfc3-4d61-81bb-df9559a566f0.png)

am I doing something wrong? 
Thanks team :) 

**Craft & Plugin Info (please complete the following information):**

- Craft Version: Craft Pro 3.7.51
- Freeform Version: 3.13.17
- Freeform Edition: pro
- Fresh Install or Upgrade: fresh

**Additional context**

![Screen Shot 2022-08-11 at 2 23 10 PM](https://user-images.githubusercontent.com/13962572/184054516-4d6b431d-de71-471e-ae2b-7aea8a4626f8.png)

**Comment by kjmartens** [comment]
Sorry for the delay @palomamtnez.

Can you see if this works instead? 🙂 

```
Event::on(
    SubmissionsService::class,
    SubmissionsService::EVENT_AFTER_SUBMIT,
    function (SubmitEvent $event) {
        $form = $event->getForm();
        $submission = $event->getSubmission();
        try {
            $myNewValue = getenv('CUSTOM_URL') . $event->getSubmission()->id;
            $submission->setFormFieldValues(['submissionUrl' => $mynewValue]);
```

**Comment by palomamtnez** [comment]
Hey @kjmartens thanks for getting back to me :)

It still has the same behaviour...
Actually, when I use that approach, then all the rest of the fields are now in blank on my confirmation email, but in the submission entry in the CMS it still shows the values prior to that.

Code:
```

use Solspace\Freeform\Events\Submissions\SubmitEvent;
use Solspace\Freeform\Services\SubmissionsService;

....


      Event::on(
            SubmissionsService::class,
            SubmissionsService::EVENT_AFTER_SUBMIT,
            function (SubmitEvent $event) {
                $form = $event->getForm();
                $submission = $event->getSubmission();
                try {
                    $submissionUrl =  getenv('CUSTOM_URL') . $submission->id;
                    $submission->setFormFieldValues(['submissionUrl' => $submissionUrl]);
```

Email:
<img width="541" alt="Screen Shot 2022-08-17 at 1 09 38 PM" src="https://user-images.githubusercontent.com/13962572/185012622-4add0be2-2427-442e-a247-52ca9f6c7004.png">

Admin panel:
<img width="755" alt="Screen Shot 2022-08-17 at 1 10 39 PM" src="https://user-images.githubusercontent.com/13962572/185012752-bcbb7991-c113-460b-8cee-b9b4b601dc6e.png">

Also, does this event, EVENT_AFTER_SUBMIT, work before the CRM integrations? 
I need to override the value of my field form, "submissionUrl", so then it can be sent in the Hubspot integration that I made via admin panel (CRM connections) 

Thanks
Paloma

**Comment by kjmartens** [comment, marked as answer]
It seems like the best possible approach here would be using the `SubmissionsService::EVENT_AFTER_SUBMIT` event.

You would have to update the `submissionUrl` field value for the `$submission` element and then re-save the element. There might be an issue while saving using elements service because of all the on-save events triggering, so it could be possible that modifying the value via database directly would be the only viable approach here, but that has to be tested.

After the data has been updated and persisted, the integrations would kick in and send the data to HubSpot as well as send out the emails.

The emails though, depend on how your email template is set up. If it uses the form data, you would also have to update the form data in the event, however, if it's using the submission data, you should be good to go.

**Comment by jeroenlammerts** [comment]
I had to run setFormFieldValues with the second param override false, and then save the submission.
`$submission->setFormFieldValues(['submissionUrl' => $submissionUrl], false);`
`Craft::$app->elements->saveElement($submission);`

Maybe add something about it to the docs?
https://docs.solspace.com/craft/freeform/v4/developer/form-events.html?#after-submitting
